Integrations/python/deephaven/Types.py

```diff
@@ -0,0 +1,229 @@
+                        type(data_type).__name__)
+    if len(t) == 2:
+        return _col_def_.fromGenericType(col_name, data_type)
+    component_type = t[2]
+    if not isinstance(component_type, _jclazz_type_):
+        raise Exception('Element at index 2 (' + str(component_type) + ') for component type has the wrong type ' +
+                        type(component_type).__name__)
+    return _col_def_.fromGenericType(col_name, data_type, component_type)
+
+@_passThrough
+def cols(ts):
+    """
+    Convert a sequence of tuples of the form ('Price', double_type)
+    or ('Prices', double_array_type, double_type)
+    to a list of ColumnDefinition objects.
+
+    :param ts: a sequence of 2 or 3 element tuples of (str, type) or (str, type, type)  specifying a column definition object.
+    :return: a list of column definition objects.
+    """
+    r = []
+    for t in ts:
+        r.append(col(t))
+    return r
+
+
+#
+# import dhtypes as dh
+#
+# data = [['tom', 10], ['nick', 15], ['juli', 14]]
+# columns = [ ('Name' : dh.string), ('Price' : dh.double) ]
+# t = dh.table_of(data, columns)
+#
+#
+@_passThrough
+def table_of(data, columns):
+    """
+    Create a Deephaven table; the API should feel familiar to users of Panda's DataFrame.
+
+    :param data: a list of rows, where each row is in turn a list of values.
+    :param columns: a list of tuples of the form (name, type) where name is the intended column name
+                    and type is the intended Deephaven column data type.
+    :return: a table with columns and data as specified in the arguments.
+    """
+
+    if columns is None or len(columns) == 0:
+        if data is not None or len(data) > 0:
+            raise Exception("when no column definitions are provided in the 'columns' argument, " +
+                            "only an empty table can be created, and no data can be specified; instead " +
+                            "got a non-empty 'data' argument with " + str(data))
+        return _table_tools_.emptyTable()
+
+    if isinstance(columns, str) or not isinstance(columns, collections.Sequence) or len(columns) < 1:
+        raise Exception("argument 'columns' needs to be a sequence with at least one element, " +
+                        "instead got " + str(columns) + " of type " + type(columns).__name__)
+    
+    if data is None or len(data) == 0:
+        col_header = None
+        for t in columns:
+            if len(t) != 2:
+                raise Exception("only two element type tuples are supported, instead got " + str(t))
+            try:
+                if col_header is None:
+                    col_header = _qst_col_header_.of(t[0], t[1])
+                else:
+                    col_header = col_header.header(t[0], t[1])
+            except Exception as e:
+                raise Exception("Could not create column definition from " + str(t)) from e
+        return _table_.of(col_header)
```

Comment:
I don't think this is valid, can't call `of` on a col_header. To create a table with zero rows from a ColumnHeader(s) you can do:

`_table_.of(_qst_newtable_.empty(col_header.tableHeader()))`, or something to that effect.

Alternatively, it's perfectly fine to not handle this as a special case, and just fall through to the general case and create Columns with no data in it.

Reply:
We could make the `NewTable.empty` method a bit more generic, and accept `Iterable<ColumnHeader>` instead of `TableHeader`, in which case we can get rid of one layer of adapting:

```python
_table_.of(_qst_newtable_.empty(col_header))
```